proposals/NNNN-allow-taskgroup-childtaskresult-type-to-be-inferred.md

```diff
@@ -0,0 +1,154 @@
+
+## Proposed solution
+
+Adding a default `ChildTaskResult.self` argument for `of childTaskResultType: ChildTaskResult.Type` will allow `withTaskGroup(of:returning:body:)` to infer the type of `ChildTaskResult` in most cases.  The currently signature of `withTaskGroup(of:returning:body:)` looks like:
+
+```swift
+public func withTaskGroup<ChildTaskResult, GroupResult>(
+    of childTaskResultType: ChildTaskResult.Type,
+    returning returnType: GroupResult.Type = GroupResult.self,
+    body: (inout TaskGroup<ChildTaskResult>) async -> GroupResult
+) async -> GroupResult where ChildTaskResult : Sendable
+```
+
+The function signature of `withThrowingTaskGroup(of:returning:body:)` is nearly identical, so only `withTaskGroup(of:returning:body:)` will be used as an example throughout this proposal.
+
+Note that the `GroupResult` generic is inferrable via the `= GroupResult.self` default argument.  This can also be applied to `ChildTaskResult` as of [SE-0326](0326-extending-multi-statement-closure-inference.md).  As in:
+
+```swift
+public func withTaskGroup<ChildTaskResult, GroupResult>(
+    of childTaskResultType: ChildTaskResult.Type = ChildTaskResult.self, // <- Updated.
+    returning returnType: GroupResult.Type = GroupResult.self,
+    body: (inout TaskGroup<ChildTaskResult>) async -> GroupResult
+) async -> GroupResult where ChildTaskResult : Sendable
+```
+
+This allows the original example above to be simplified:
+
+```swift
+let s = withTaskGroup { group in
+  group.addTask { someVoidFunction() }
+  group.addTask { someVoidFunction() }
+  group.addTask { someVoidFunction() }
+
+  return "Hello, world!"
+}
+```
+
+In the above snippet, `ChildTaskResult` is inferred as `Void` and `GroupResult` is inferred as `String`.  Not needing to specify the generics explicitly will simplify the API design for these functions and make it easier for new users of these APIs, as it can currently be confusing to understand the differences between `ChildTaskResult` and `GroupResult` (especially when one or both of those is `Void`).
+
+## Detailed design
+
+Because type inference is top-down, it relies on the first statement that uses `group` to infer determine the generic for `ChildTaskResult`.  Therefore, it is possible to get a compiler error by creating a task group where the first use of `group` does not use `addTask(...)`, like so:
+
+```swift
+// Expect `ChildTaskResult` to be `Void`...
+withTaskGroup { group in // Generic parameter 'ChildTaskResult' could not be inferred
+    // Since `addTask(...)` wasn't the first statement, this fails to compile.
+    group.cancelAll()
+    group.addTask { 42 }
+}
+```
+
+This can be fixed by going back to specifying the generic like before:
+
+```swift
+// Expect `ChildTaskResult` to be `Int`...
+withTaskGroup(of: Int.self) { group in
+    group.cancelAll()
+    group.addTask { 42 }
```

Comment:
I think it would be beneficial to have some more realistic examples. For example, there are small-but-realistic examples at https://www.avanderlee.com/concurrency/task-groups-in-swift/ and https://www.kodeco.com/books/modern-concurrency-in-swift/v1.0/chapters/7-concurrent-code-with-taskgroup that you could reference (with attribution!) and will give reviewers a better feel for how this impacts use of the feature.

Reply:
I have updated the examples to use more real-world scenarios.  Thank you for the suggestion.  👍